vortex-datafusion/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+    ) -> DFResult<Arc<dyn ExecutionPlan>> {
+        if !filters.is_empty() {
+            return exec_err!("vortex does not support filter pushdown");
+        }
+
+        let partitioning = if let Ok(chunked_array) = ChunkedArray::try_from(&self.array) {
+            Partitioning::RoundRobinBatch(chunked_array.nchunks())
+        } else {
+            Partitioning::UnknownPartitioning(1)
+        };
+
+        let plan_properties = PlanProperties::new(
+            EquivalenceProperties::new(self.schema_ref.clone()),
+            partitioning,
+            ExecutionMode::Bounded,
+        );
+
+        Ok(Arc::new(VortexMemoryExec {
+            array: self.array.clone(),
+            projection: projection.cloned(),
+            plan_properties,
+        }))
+    }
+
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> DFResult<Vec<TableProviderFilterPushDown>> {
+        // TODO(aduffy): add support for filter pushdown
+        Ok(filters
+            .iter()
+            .map(|_| TableProviderFilterPushDown::Unsupported)
+            .collect())
+    }
+}
+
+/// Physical plan node for scans against an in-memory, possibly chunked Vortex Array.
+#[derive(Debug, Clone)]
+struct VortexMemoryExec {
+    array: Array,
+    projection: Option<Vec<usize>>,
+    plan_properties: PlanProperties,
+}
+
+impl DisplayAs for VortexMemoryExec {
+    fn fmt_as(&self, _display_type: DisplayFormatType, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl VortexMemoryExec {
+    /// Read a single array chunk from the source as a RecordBatch.
+    fn execute_single_chunk(
+        array: Array,
+        projection: &Option<Vec<usize>>,
+        _context: Arc<TaskContext>,
+    ) -> DFResult<SendableRecordBatchStream> {
+        let data = array
+            .flatten()
+            .map_err(|vortex_error| DataFusionError::Execution(format!("{}", vortex_error)))?
+            .into_array();
+
+        // Construct the RecordBatch by flattening each struct field and transmuting to an ArrayRef.
+        let struct_array = StructArray::try_from(data)
+            .map_err(|_| exec_datafusion_err!("top-level array must be Struct encoding"))?;
+
+        let field_order = if let Some(projection) = projection {
+            projection.clone()
+        } else {
+            (0..struct_array.names().len()).collect()
+        };
+
+        let mut record_batch = Vec::with_capacity(field_order.len());
+        let mut record_batch_fields = Vec::with_capacity(field_order.len());
+
+        for field_idx in field_order {
+            let field_array = struct_array
+                .field(field_idx)
+                .expect("field array must be present")
+                .flatten()
+                .map_err(|_| exec_datafusion_err!("failed to flatten struct field array"))?;
+
+            let array_ref = match field_array {
+                Flattened::Null(a) => AsArrowArray::as_arrow(&a).unwrap(),
```

Comment:
`AsArrowArray::as_arrow(&field_array.into_array()).expect("Flat arrays must convert to arrow")`

Or maybe just add an `as_arrow` function to Flattened? Then can unwrap the error inside there.

Reply:
will add to Flattened

Reply:
done